client/tinychain/ml/linalg.py

```diff
@@ -273,3 +277,235 @@ def svd(matrix: Tensor) -> Tuple:
+                cxt.row_j = U[j, l:]
+                cxt.update = cxt.row_j.write(cxt.row_j + e[l:] * dot(cxt.row_j, outer_cxt.row_i))
+                return After(cxt.update, Map(j=j + 1))
+
+            @closure(txn.m)
+            @post_op
+            def cond(j: UInt) -> Bool:
+                return j < txn.m
+
+            cxt.U_i_l = U[i, l]
+            cxt.new_row = (cxt.row_i / scale).copy()
+            cxt.descale = cxt.row_i.write(cxt.new_row)
+
+            cxt.s_init = Float(After(cxt.descale, dot(cxt.row_i, cxt.row_i)))
+            cxt.f = Float(After(cxt.descale, U[i, l]))
+            cxt.g_init = -sign_like(cxt.s_init ** 0.5, cxt.f)
+
+            cxt.h = After(cxt.U_i_l.write(cxt.f - cxt.g_init), (cxt.f * cxt.g_init) - cxt.s_init)
+            cxt.s_final = After(e[l:].write(cxt.row_i / cxt.h), cxt.s_init)
+            cxt.g_final = After(While(cond, step, Map(j=l)), cxt.g_init)
+
+            cxt.rescale = After(cxt.s_final, cxt.row_i.write(cxt.row_i * scale))
+            return After(cxt.rescale, cxt.g_final)
+
+        cxt.update_cols = update_cols
+        cxt.update_rows = update_rows
+
+        cxt.update_e = e[i].write(scale * g)
+        cxt.l = UInt(After(cxt.update_e, i + 1))
+        cxt.scale_init = Float(If(i < txn.m, dot(U[i:, i], U[i:, i]), scale))
+        cxt.g_init = Float(If(
+            Bool(i < txn.m).logical_and(cxt.scale_init.abs() > EPS),
+            cxt.update_cols(i=i, l=cxt.l, scale=cxt.scale_init),
+            0.0))
+
+        cxt.scale_new = Float(If(
+            Bool(i < txn.m).logical_and(i != txn.n - 1),
+            dot(U[i, cxt.l:], U[i, cxt.l:]),
+            scale))
+
+        cxt.scale_final = Float(After(W[i].write(cxt.scale_init * cxt.g_init), cxt.scale_new))
+
+        g = If(
+            Bool(i < txn.m).logical_and(i != txn.n - 1).logical_and(cxt.scale_init > EPS),
+            cxt.update_rows(l=cxt.l, scale=cxt.scale_final),
+            0.0)
+
+        return Map(i=i + 1, scale=cxt.scale_final, g=g, U=U, W=W, e=e)
+
+    @closure(txn.n)
+    @post_op
+    def cond(i: UInt) -> Bool:
+        return i < txn.n
+
+    return Map(While(cond, step, Map(i=0, scale=0.0, g=0.0, U=U, W=W, e=e)))
+
+
+@post_op
+def rht(txn, U: Tensor, V: Tensor, e: Tensor) -> Map:
+    """Accumulation of right hand transformations (rht)"""
+
+    txn.shape = U.shape
+    txn.m, txn.n = [UInt(dim) for dim in txn.shape.unpack(2)]
+    
+    @closure(U, V, e, txn.m, txn.n)
+    @post_op
+    def step(cxt, i: UInt, l: UInt, g: F32) -> Map:
+        
+        @closure(U, V, e, txn.m, txn.n)
+        @post_op
+        def update_cols_wo_last(cxt, i: UInt, l: UInt, g: F32) -> F32:
+
+            @closure(U, V, e, txn.m, txn.n)
+            @post_op
+            def update_cols(cxt, i: UInt, l: UInt, g: F32) -> F32:
+                
+                @closure(U, V, e, txn.m, txn.n)
+                @post_op
+                def step(j: UInt, l: UInt, i: UInt) -> Map:
+                    cxt.col_j = V[l:, j]
```

Comment:
note that you're assigning here to the outer scope, I think you mean to include a `cxt` param like `def step(cxt, ...)`